handle.go

```diff
@@ -0,0 +1,153 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build (linux || darwin) && (amd64 || arm64)
+
+package waf
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+
+	emptyfree "github.com/DataDog/go-libddwaf/internal/empty-free"
+	"go.uber.org/atomic"
+)
+
+// Handle represents an instance of the WAF for a given ruleset.
+type Handle struct {
+	cHandle        wafHandle
+	mutex          sync.RWMutex
+	contextCounter *atomic.Uint32
+	rulesetInfo    RulesetInfo
+}
+
+// NewHandle takes rules and the obfuscator config and return a new WAF to work with. The order of action is the following:
+// - Open the WAF library libddwaf.so using dlopen
+// - Encode the rule free form object as a ddwaf_object so send them to the WAF
+// - Create a ddwaf_config object and fill the values
+// - Run ddwaf_init to get a handle from the waf
+// - Check for errors and streamline the ddwaf_ruleset_info returned
+func NewHandle(rules any, keyObfuscatorRegex string, valueObfuscatorRegex string) (*Handle, error) {
+
+	if err := InitWaf(); err != nil {
+		return nil, err
+	}
+
+	if reflect.ValueOf(rules).Type() == reflect.TypeOf([]byte(nil)) {
+		return nil, errors.New("cannot encode byte array as top-level rules object")
+	}
+
+	encoder := newMaxEncoder()
+	obj, err := encoder.Encode(rules)
+	if err != nil {
+		return nil, fmt.Errorf("could not encode the WAF ruleset into a WAF object: %w", err)
+	}
+
+	defer encoder.allocator.KeepAlive()
+
+	config := newConfig(&encoder.allocator, keyObfuscatorRegex, valueObfuscatorRegex)
+	cRulesetInfo := new(wafRulesetInfo)
+
+	cHandle := wafLib.wafInit(obj, config, cRulesetInfo)
+	if cHandle == 0 {
+		return nil, errors.New("could not instanciate the WAF")
+	}
+
+	defer wafLib.wafRulesetInfoFree(cRulesetInfo)
+
+	errorsMap, err := decodeErrors(&cRulesetInfo.errors)
+	if err != nil { // Something is very wrong
+		return nil, fmt.Errorf("could not decode the WAF ruleset errors: %w", err)
+	}
+
+	return &Handle{
+		cHandle:        cHandle,
+		contextCounter: atomic.NewUint32(1), // We count the handle itself in the counter
+		rulesetInfo: RulesetInfo{
+			Loaded:  cRulesetInfo.loaded,
+			Failed:  cRulesetInfo.failed,
+			Errors:  errorsMap,
+			Version: gostring(cRulesetInfo.version),
+		},
+	}, nil
+}
+
+// NewContext a new WAF context and increase the number of references to the WAF
+// handle. A nil value is returned when the WAF handle can no longer be used
+// or the WAF context couldn't be created.
+func (handle *Handle) NewContext() *Context {
+	// Handle has been released
+	if handle.contextCounter.Load() == 0 {
+		return nil
+	}
```

Comment:
The previous CAS implementation of the handle ref counter is still needed to avoid a race between the context creations and their WAF handle closing. `Load() == 0` isn't enough, and can become true right after it was false in a call to NewContext.

Reply:
I added a test covering stressing the ref-counter in https://github.com/DataDog/go-libddwaf/pull/15/commits/8bb0d5b0310c6adebac3f7da214c324f9d5fb7a8 

It works like a charm and finds the issue :-)

Reply:
It's working :)